### PR TITLE
multi-machine kvstore

### PR DESCRIPTION
**Pull request by mli**
passed unit test, see

tests/python/distributed/test_kvstore.py

### CODE REVIEW

File: include/mxnet/kvstore.h [hunk 1]
```diff
@@ -106,40 +123,153 @@ class KVStore {
...
+   *
+   * Send a command to all server nodes, which will make each server node run
+   * \a controller
+   *
+   * This function returns after the command has been executed in all server nodes
+   *
+   * \param head the head of the command
+   * \param body the body of the command
+   */
+  virtual void SendCommandToServers(int head, const std::string& body) { }
```
Comment (tqchen) [inline include/mxnet/kvstore.h]: The head parameter is unclear

File: python/mxnet/kvstore.py [hunk 2]
```diff
@@ -233,6 +245,131 @@ def set_updater(self, updater):
...
+        -------
+        size :int
+            The number of worker nodes
+        """
+        size = ctypes.c_int()
+        check_call(_LIB.MXKVStoreGetGroupSize(self.handle, ctypes.byref(size)))
+        return size.value
+
+
+    def barrier(self):
```
Comment (tqchen) [inline python/mxnet/kvstore.py]: Do we really need to expose this?

File: python/mxnet/kvstore.py [hunk 3]
```diff
@@ -233,6 +245,131 @@ def set_updater(self, updater):
...
+
+        But note that, this functions only blocks the main thread of workers
+        until all of them are reached this point. It doesn't guarantee that all
+        operations issued before are actually finished, such as \ref Push and
+        \ref Pull. In that case, we need to call \ref Wait or \ref WaitAll
+        """
+        check_call(_LIB.MXKVStoreBarrier(self.handle))
+
+    def wait(self, key):
+        """Wait until all pushes and pulls issued on each key have been finished
```
Comment (tqchen) [inline python/mxnet/kvstore.py]: This seems not necessary?

File: python/mxnet/kvstore.py (hunk 2 above)
Comment (tqchen) [inline python/mxnet/kvstore.py]: Consider put this as internal function _barrier

File: python/mxnet/kvstore.py [hunk 4]
```diff
@@ -233,6 +245,131 @@ def set_updater(self, updater):
...
+                assert(isinstance(k, int))
+            ckeys = c_array(ctypes.c_int, key)
+        check_call(_LIB.MXKVStoreWait(self.handle, mx_uint(len(ckeys)), ckeys))
+
+    def wait_all(self):
+        """Wait until all pushes and pulls issued before have been finished
+        """
+        check_call(_LIB.MXKVStoreWaitAll(self.handle))
+
+    def send_command_to_servers(self, head, body):
```
Comment (tqchen) [inline python/mxnet/kvstore.py]: make it private

File: python/mxnet/optimizer.py [hunk 5]
```diff
@@ -122,10 +138,34 @@ def create(name, rescale_grad=1, **kwargs):
...
     """
     if name == 'sgd' or name == 'SGD':
         return SGD(rescale_grad=rescale_grad, **kwargs)
+    if name == 'test':
+        return Test(rescale_grad=rescale_grad)
     else:
         raise ValueError('Cannot find optimizer %s' % name)
+
+
+def optimizer_clossure(optimizer):
```
Comment (tqchen) [inline python/mxnet/optimizer.py]: name is not obvious, maybe kvstore_updater(optimizer)

File: src/kvstore/mxnet_node.h [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * \file mxnet_node.h
+ * \brief implement mxnet nodes
+ */
+#ifndef MXNET_KVSTORE_MXNET_NODE_H_
+#define MXNET_KVSTORE_MXNET_NODE_H_
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+#include <memory>
+#include <functional>
```
Comment (tqchen) [inline src/kvstore/mxnet_node.h]: maybe mxnet_ps_node.h

File: src/kvstore/mxnet_node.h [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+#include <vector>
+#include "ps.h"
```
Comment (tqchen) [inline src/kvstore/mxnet_node.h]: if it is in include use <ps.h>

File: src/kvstore/mxnet_node.h [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+#include "mxnet/kvstore.h"
```
Comment (tqchen) [inline src/kvstore/mxnet_node.h]: <mxnet/kvstore.h>

File: python/mxnet/kvstore_server.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
+# coding: utf-8
+""" a server node for the key value store """
+from __future__ import absolute_import
+import ctypes
+import sys
+from .base import _LIB, check_call
+from .kvstore import create
+
+class KVStoreServer(object):
```
Comment (tqchen) [inline python/mxnet/kvstore_server.py]: consider move it back into kvstore.py, Also we should not expose server to user

File: include/mxnet/kvstore.h (hunk 1 above)
Comment (tqchen) [inline include/mxnet/kvstore.h]: add more docstring, or simply add a enum type about KVStoreCmdType

File: include/mxnet/kvstore.h [hunk 10]
```diff
@@ -106,40 +123,153 @@ class KVStore {
...
+   * \return whether or not this process is a scheduler node.
    *
-   * \param batch true for batch, false for online
-   * \param updt user-defined updater, default is assign
+   * Always returns false when type == "local"
    */
-  virtual void set_updater(Updater updater) = 0;
+  static bool IsSchedulerNode() {
+    char* role_str = getenv("DMLC_ROLE");
+    return (role_str != NULL) && (!strcmp(role_str, "scheduler"));
```
Comment (tqchen) [inline include/mxnet/kvstore.h]: nullptr

**Comment by tqchen** [comment]
current kv seems was not python3 compatible.

**Comment by tqchen** [comment]
Fix pylint style as https://travis-ci.org/dmlc/mxnet/jobs/84239796

**Comment by tqchen** [comment]
You can do it locally by make lint

**Comment by tqchen** [comment]
Also need to rebase as http://mxnet.readthedocs.org/en/latest/contribute.html#submit-a-pull-request

File: include/mxnet/kvstore.h (hunk 10 above)
Comment (mli) [inline include/mxnet/kvstore.h]: will be set as null in default

Comment (tqchen) [inline include/mxnet/kvstore.h]: I mean use nullptr to replace NULL

**Pull request merged**